[WV-1199] Add dataLayer to 'See how to turn your values into voting decisions!" on Ballot Page [MERGE READY]

dboudagian opened the pull request:
### What github.com/wevote/WebApp/issues does this fix?

Fixes WV-1199

### Changes included this pull request?
- Added dataLayer tracking for hideHowItWorksWizard in HowItWorksWizard.jsx
- Added dataLayer tracking for openHowItWorksModal in CompleteYourProfile2024.jsx
- Added dataLayer tracking for openPersonalizedScoreIntroModal in CompleteYourProfile2024.jsx
- Added dataLayer tracking for toggleShowSignInModal in CompleteYourProfile2024.jsx
- Tracks all user interactions with the Complete Your Profile wizard on Ballot page
- Fixed all ESLint errors and warnings in both components

dboudagian committed:
[WV-1199] Add dataLayer to 'See how to turn your values into voting decisions!' on Ballot Page

- Added dataLayer tracking for hideHowItWorksWizard in HowItWorksWizard.jsx
- Added dataLayer tracking for openHowItWorksModal in CompleteYourProfile2024.jsx
- Added dataLayer tracking for openPersonalizedScoreIntroModal in CompleteYourProfile2024.jsx
- Added dataLayer tracking for toggleShowSignInModal in CompleteYourProfile2024.jsx
- Tracks all user interactions with the Complete Your Profile wizard on Ballot page
- Fixed all ESLint errors and warnings in both components

dboudagian commented:
@DaleMcGrew This PR implements the dataLayer tracking for the "See how to turn your values into voting decisions!" feature as requested in WV-1199. Ready for review!

dboudagian committed:
[WV-1199] Fix build issues

DaleMcGrew commented on src/js/components/CompleteYourProfile/CompleteYourProfile2024.jsx:
```diff
@@ -221,6 +286,45 @@ class CompleteYourProfile2024 extends Component {
+
+      const dataLayerObject = {
+        actionDetails: {
+          actionType: 'openModal',
```
Adjust actionType to dynamically be either 'openModal' or 'closeModal'.

DaleMcGrew commented on src/js/components/CompleteYourProfile/CompleteYourProfile2024.jsx:
```diff
@@ -183,12 +185,75 @@ class CompleteYourProfile2024 extends Component {
+      },
+      event: 'action',
+      destinationDetails: {
+        destinationPageName: currentPage.pageName,
```
I think for destinationPageName we want 'HowItWorksWizard'.

DaleMcGrew commented on src/js/components/CompleteYourProfile/HowItWorksWizard.jsx:
```diff
@@ -14,8 +17,40 @@ const HowItWorksWizard = ({ steps, activeStep }) => {
+      },
+      event: 'action',
+      pageDetails: {
+        PageName: currentPage.pageName,
```
I think for pageName we want 'HowItWorksWizard'.

DaleMcGrew commented on src/js/components/CompleteYourProfile/HowItWorksWizard.jsx:
```diff
@@ -14,8 +17,40 @@ const HowItWorksWizard = ({ steps, activeStep }) => {
+    const dataLayerObject = {
+      actionDetails: {
+        actionType: 'close',
+        buttonId: 'CloseHowItWorksWizard',
```
Please add this same 'CloseHowItWorksWizard' id to <HowItWorksCrossIconContainer so the QA Automation team can also trigger the onClick event.
I think it would be more elegant to pass in the buttonId through the hideHowItWorksWizard function call to avoid a future mismatch if the id is changed on <HowItWorksCrossIconContainer and doesn't get changed here.

DaleMcGrew commented on src/js/components/CompleteYourProfile/CompleteYourProfile2024.jsx:
```diff
@@ -221,6 +286,45 @@ class CompleteYourProfile2024 extends Component {
+        event: 'action',
+        destinationDetails: {
+          destinationPageName: currentPage.pageName,
+          destinationPageType: 'SignInModal',
```
I think we want this instead:
```
        destinationDetails: {
          destinationPageName: 'SignInModal',
          destinationPageType: currentPage.pageType,
```

DaleMcGrew commented on src/js/components/CompleteYourProfile/CompleteYourProfile2024.jsx:
```diff
@@ -183,12 +185,75 @@ class CompleteYourProfile2024 extends Component {
+      event: 'action',
+      destinationDetails: {
+        destinationPageName: currentPage.pageName,
+        destinationPageType: 'PersonalizedScoreIntroModal',
```
I think this should be:
```
      destinationDetails: {
        destinationPageName: 'PersonalizedScoreIntroModal',
        destinationPageType: currentPage.pageType,
```

DaleMcGrew reviewed (changes requested):
Great work @dboudagian! Some change requests below.

DaleMcGrew commented:
I just realized that we should send a dataLayer when the src/js/components/CompleteYourProfile/CompleteYourProfile2024.jsx component is loaded. The reason is that when we have run analysis about how many people clicked on any of the steps, we need to be able to say how many people were offered these steps.

dboudagian committed:
feat: Add dataLayer tracking to CompleteYourProfile wizard

- Add impression tracking when CompleteYourProfile2024 component loads
- Add close tracking to HowItWorksWizard with dynamic buttonId
- Add id='CloseHowItWorksWizard' to close button for QA automation
- Make buttonId parameter dynamic to prevent future mismatches
- Fix ESLint object-shorthand warning

DaleMcGrew commented on src/js/components/CompleteYourProfile/HowItWorksWizard.jsx:
```diff
@@ -1,21 +1,56 @@
+      event: 'action',
+      pageDetails: {
+        PageName: 'HowItWorksWizard',
+        PageType: currentPage.pageType,
```
You can fix in follow-up pull request, but these should start with a lowercase letter:
pageName
pageType

DaleMcGrew commented on src/js/components/CompleteYourProfile/HowItWorksWizard.jsx:
```diff
@@ -27,7 +62,10 @@ const HowItWorksWizard = ({ steps, activeStep }) => {
         </p>
-        <HowItWorksCrossIconContainer onClick={hideHowItWorksWizard}>
+        <HowItWorksCrossIconContainer
+          id="CloseHowItWorksWizard"
```
I prefer our ids start with lower case letter, like this: closeHowItWorksWizard

DaleMcGrew reviewed (commented):
Great work! You can weave my comments into your next pull request. 👍
